File: java/scf-core/src/main/java/org/mydotey/scf/DefaultProperty.java
package org.mydotey.scf;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.function.Consumer;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * @author koqizhao
 *
 * May 16, 2018
 */
public class DefaultProperty<K, V> implements Property<K, V> {

    private static final Logger LOGGER = LoggerFactory.getLogger(DefaultProperty.class);

    private PropertyConfig<K, V> _config;
    private volatile V _value;
    private volatile List<Consumer<Property<K, V>>> _changeListeners;

    public DefaultProperty(PropertyConfig<K, V> config, V value) {
        Objects.requireNonNull(config, "config is null");

        _config = config;
        _value = value;
    }

    @Override
    public PropertyConfig<K, V> getConfig() {
        return _config;
    }

    @Override
    public V getValue() {
        return _value;
    }

    public void setValue(V value) {
        _value = value;
    }

    @Override
    public synchronized void addChangeListener(Consumer<Property<K, V>> changeListener) {
        Objects.requireNonNull("changeListener", "changeListener is null");

        if (_changeListeners == null)
            _changeListeners = new ArrayList<>();
        _changeListeners.add(changeListener);
    }

    public synchronized void raiseChangeEvent() {
        if (_changeListeners == null)
            return;

        _changeListeners.forEach(l -> {
            try {
                l.accept(DefaultProperty.this);
            } catch (Exception e) {
                LOGGER.error("property change listener failed to run", e);
            }
        });
    }

    @Override
    public String toString() {
        return String.format("%s { config: %s, value: %s, changeListeners: %s }", getClass().getSimpleName(), _config,
                _value, _changeListeners);
    }

}

File: dotnet/src/MyDotey.SCF/DefaultProperty.cs
using System;
using System.Collections.Generic;
using System.Threading;

using NLog;

namespace MyDotey.SCF
{
    /**
     * @author koqizhao
     *
     * May 16, 2018
     */
    public class DefaultProperty<K, V> : IProperty<K, V>
    {
        private static Logger Logger = LogManager.GetCurrentClassLogger(typeof(DefaultProperty<,>));

        private PropertyConfig<K, V> _config;
        private volatile object _value;
        private volatile IConfigurationSource _source;
        private volatile List<EventHandler<IPropertyChangeEvent<K, V>>> _changeListeners;

        public DefaultProperty(PropertyConfig<K, V> config, V value, IConfigurationSource source)
        {
            if (config == null)
                throw new ArgumentNullException("config is null");

            _config = config;
            _value = value;
            _source = source;
        }

        IPropertyConfig IProperty.Config { get { return Config; } }

        public virtual PropertyConfig<K, V> Config { get { return _config; } }

        object IProperty.Value { get { return Value; } }

        public virtual V Value { get { return (V)_value; } }

        public virtual IConfigurationSource Source { get { return _source; } }

        protected internal virtual void Update(object value, IConfigurationSource source)
        {
            _value = (V)value;
            _source = source;
        }

        public virtual event EventHandler<IPropertyChangeEvent<K, V>> OnChange
        {
            add
            {
                if (value == null)
                    throw new ArgumentNullException("changeListener is null");

                lock (this)
                {
                    if (_changeListeners == null)
                        _changeListeners = new List<EventHandler<IPropertyChangeEvent<K, V>>>();
                    _changeListeners.Add(value);
                }
            }

            remove
            {
                throw new NotSupportedException("remove is not supported");
            }
        }

        protected virtual void RaiseChangeEvent(IPropertyChangeEvent<K, V> @event)
        {
            if (_changeListeners == null)
                return;

            lock (this)
            {
                _changeListeners.ForEach(l =>
                {
                    try
                    {
                        l(this, @event);
                    }
                    catch (Exception e)
                    {
                        Logger.Error(e, "property change listener failed to run");
                    }
                });
            }
        }

        public override string ToString()
        {
            return string.Format("{0} {{ config: {1}, value: {2}, source: {3}, changeListeners: [ {4} ] }}",
                GetType().Name, _config, _value, _source == null ? null : _source.Config.Name,
                _changeListeners == null ? null : string.Join(", ", _changeListeners));
        }
    }
}